### ISSUE TITLE
Impossible case is possible

### ISSUE THREAD

**Issue by Jaxan** [idris-lang/Idris-dev#3846]
When case splitting the following function I get impossible cases:

```
plusOne : (answer : Vect (len + 1) x) -> Vect (S len) x
plusOne answer = ?plusOne_rhs
```
Turns into:
```
plusOne : (answer : Vect (len + 1) x) -> Vect (S len) x
plusOne [] impossible
plusOne (_ :: _) impossible
```

Then, when you compile, you get:
```
plusOne [] is a valid case
```

Idris Version 1.0-git:64c5f702

I found a similar issue #3663, which is closed. But I'm using the current development version, so a similar issue is still there.

**Comment by mrayva** [comment]
г fxzzj

**Comment by pandemonium** [comment]
My, albeit limited, experience in Idris tells me that this is because the typer is unable to unify `len + 1` with `S len`, so you need to prove that they are the same. (I men: they are, but something with the order of the terms means that they are not guaranteed to be because the operator `+` is not reflexive unless you prove that [to Idris.])

**Comment by Jaxan** [comment]
I am not sure the return type is very relevant for case splitting here. In any case, I see the following inconsistency:

* The case splitter makes an under approximation: It sees `Vect (len + 1)` and declares both `Z` and `S k` as impossible.
* The type checker actually makes an over approximation: It even considers the case `Z` possible.

**Comment by edwinb** [comment]
It's because neither case type checks, and the case splitter isn't very clever about realising that it's not the input being impossible, but rather that the particular cases don't obviously unify. Probably better behaviour here would be to not split (and maybe give a message saying why it didn't).

**Comment by Jaxan** [comment]
Let me give some background: I was working my way through the book (which is really nice, btw!) and this was just something I tried along the way. I tried it way before I got to the second part, so my only hope was to try the automatic way: ctrl+alt+a and ctrl+alt+c for every function ;-).
What I found rather surprising is that the case splitter and the type checker don't agree. I now know there's ways to do it, and indeed don't involve splitting at all.
